# Assignment 3 - Text Analysis
An explanation this assignment could be found in the .pdf explanation document


## Materials to review for this assignment
<h4>From Moodle:</h4> 
<h5><u>Review the notebooks regarding the following python topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Working with strings</b> (tutorial notebook)<br/>
&#x2714; <b>Text Analysis</b> (tutorial notebook)<br/>
&#x2714; <b>Hebrew text analysis tools (tokenizer, wordnet)</b> (moodle example)<br/>
&#x2714; <b>(brief review) All previous notebooks</b><br/>
</div> 
<h5><u>Review the presentations regarding the following topics</u>:</h5>
<div class="alert alert-info">
&#x2714; <b>Text Analysis</b> (lecture presentation)<br/>
&#x2714; <b>(brief review) All other presentations</b><br/>
</div>

## Personal Details:

In [2]:
# Details Student 1:
# Yahav Sharabi 207227943
# Details Student 2:
# Esti Rabino 206080954

## Preceding Step - import modules (packages)
This step is necessary in order to use external modules (packages). <br/>

In [3]:
# --------------------------------------
import pandas as pd
import numpy as np
# --------------------------------------


# --------------------------------------
# ------------- visualizations:
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
# --------------------------------------


# ---------------------------------------
import sklearn
from sklearn import preprocessing, metrics, pipeline, model_selection, feature_extraction 
from sklearn import naive_bayes, linear_model, svm, neural_network, neighbors, tree
from sklearn import decomposition, cluster

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV 
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.metrics import precision_score, recall_score, f1_score
from sklearn.metrics import mean_squared_error, r2_score, silhouette_score
from sklearn.preprocessing import MinMaxScaler, StandardScaler, LabelEncoder

from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import Perceptron, SGDClassifier
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
# ---------------------------------------


# ----------------- output and visualizations: 
import warnings
from sklearn.exceptions import ConvergenceWarning
warnings.simplefilter("ignore")
warnings.simplefilter(action='ignore', category=FutureWarning)
warnings.simplefilter("ignore", category=ConvergenceWarning)
# show several prints in one cell. This will allow us to condence every trick in one cell.
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
%matplotlib inline
pd.pandas.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
# ---------------------------------------

### Text analysis and String manipulation imports:

In [4]:
# --------------------------------------
# --------- Text analysis and Hebrew text analysis imports:
# vectorizers:
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# regular expressions:
import re
# --------------------------------------

### (optional) Hebrew text analysis - WordNet (for Hebrew)
Note: the WordNet is not a must

#### (optional) Only if you didn't install Wordnet (for Hebrew) use:

In [5]:
# word net installation:

# unmark if you want to use and need to install
#!pip install wn
#!python -m wn download omw-he:1.4

In [6]:
# word net import:

# unmark if you want to use:
import wn

### (optional) Hebrew text analysis - hebrew_tokenizer (Tokenizer for Hebrew)
Note: the hebrew_tokenizer is not a must

#### (optional) Only if you didn't install hebrew_tokenizer use:

In [7]:
# Hebrew tokenizer installation:

# unmark if you want to use and need to install:
#!pip install hebrew_tokenizer

In [8]:
# Hebrew tokenizer import:

# unmark if you want to use:
import hebrew_tokenizer as ht

C:\Users\esti.rabino\Downloads\assignment3-text_analysis\assignment3-text_analysis


### Reading input files
Reading input files for train annotated corpus (raw text data) corpus and for the test corpus

In [9]:
train_filename = 'annotated_corpus_for_train.csv'
test_filename  = 'corpus_for_test.csv'
df_train = pd.read_csv(train_filename, index_col=None, encoding='utf-8')
df_test  = pd.read_csv(test_filename, index_col=None, encoding='utf-8')

In [10]:
df_train.head(8)
df_train.shape

story gender
0  כשחבר הזמין אותי לחול, לא באמת חשבתי שזה יקרה,...      m
1  לפני שהתגייסתי לצבא עשיתי כל מני מיונים ליחידו...      m
2  מאז שהתחילו הלימודים חלומו של כל סטודנט זה הפנ...      f
3  כשהייתי ילד, מטוסים היה הדבר שהכי ריתק אותי. ב...      m
4  ‏הייתי מדריכה בכפר נוער ומתאם הכפר היינו צריכי...      f
5  לפני כ3 חודשים טסתי לרומא למשך שבוע. טסתי במטו...      f
6  אני כבר שנתיים נשוי והשנה אני ואישתי סוף סוף י...      m
7  השנה התחלנו שיפוץ בדירה שלנו בתל אביב. הדירה ה...      f

(753, 2)

In [11]:
df_test.head(3)
df_test.shape

test_example_id                                              story
0                0  כל קיץ אני והמשפחה נוסעים לארצות הברית לוס אנג...
1                1  הגעתי לשירות המדינה אחרי שנתיים כפעיל בתנועת "...
2                2  אחת האהבות הגדולות שלי אלו הכלבים שלי ושל אישת...

(323, 2)

### Your implementation:
Write your code solution in the following code-cells

In [12]:
def reg_clean_text(text):
    text = re.sub(r'\W', ' ', text)
    text = re.sub(r'\d', ' ', text)
    text = re.sub(r'\s\w\s', ' ', text)
    return text

In [13]:
def clean_and_process_text(df):
    processed_df = df.copy()
    processed_df['story'] = processed_df['story'].apply(reg_clean_text)
    return processed_df

In [14]:
def clean_df():
    processed_df_train = clean_and_process_text(df_train.copy())
    processed_df_test = clean_and_process_text(df_test.copy())
    X_train = processed_df_train['story']
    y_train = processed_df_train['gender']
    x_test = processed_df_test['story']
    return X_train, y_train, x_test

In [15]:
def tokenizer_vect(hebrew_text):
    token = ht.tokenize(hebrew_text)
    tokens = list( filter( lambda x : ( x[0] == 'HEBREW') , token))
    tokens = [tuple[1] for tuple in tokens]
    return tokens

In [16]:
x_train , y_train , x_test = clean_df()


<font size="5">Models:</font>

<font size="4">1. KNN</font>

In [17]:
# Create a pipeline for text classification 
pipeline = Pipeline([
    ('vect', CountVectorizer(ngram_range=(1, 6))),
    ('clf', KNeighborsClassifier())
])

#Set up a list of options for adjusting parameters when searching for the best values in hyperparameter tuning
param_grid = {
    'vect__ngram_range': [(1, 1), (1, 2)],
    'clf__n_neighbors': [3, 5, 7],
    'clf__weights': ['uniform', 'distance'],
    'clf__p': [1, 2]
}

#creating a grid search object and trying different settings for our text classification model
#in order to find the best combination of settings
grid_search = GridSearchCV(
    estimator=pipeline,
    param_grid=param_grid,
    cv=10,
    scoring='f1_macro',
    n_jobs=-1 
)
grid_search.fit(x_train, y_train)

#Create paramter that holds the result for the tested model
KNN_model = grid_search.best_estimator_

#Calculate the F1-score
f1_scores = cross_val_score(
    estimator=grid_search.best_estimator_,
    X=x_train,
    y=y_train,
    scoring='f1_macro',
    cv=10,
    n_jobs=-1
)
f1_average = f1_scores.mean()


print("F1 average score:", f1_average)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('vect',
                                        CountVectorizer(ngram_range=(1, 6))),
                                       ('clf', KNeighborsClassifier())]),
             n_jobs=-1,
             param_grid={'clf__n_neighbors': [3, 5, 7], 'clf__p': [1, 2],
                         'clf__weights': ['uniform', 'distance'],
                         'vect__ngram_range': [(1, 1), (1, 2)]},
             scoring='f1_macro')

F1 average score: 0.5441303518258724


<font size="4">2. LinearSVC:</font>


In [18]:
# Create a pipeline
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(tokenizer=tokenizer_vect)),
    ('clf', LinearSVC())
])


params = {
    'tfidf__ngram_range': [(1, 1), (1, 2)],
    'tfidf__use_idf': (True, False),
    'tfidf__max_df': (0.5, 0.75, 1.0),
    'tfidf__min_df': (1, 2),
    'clf__C': (0.1, 1, 10),
    'clf__penalty': ('l1', 'l2'),
    'clf__loss': ('hinge', 'squared_hinge'),
    'clf__dual': (True, False)
}

grid_search = GridSearchCV(
    estimator=pipeline,
    param_grid=params,
    cv=10,
    scoring='f1_macro',  
    n_jobs=-1  
)
grid_search.fit(x_train, y_train)

linearSVC_model = grid_search.best_estimator_

#Calculate the F1-score
f1_scores = cross_val_score(
    estimator=grid_search.best_estimator_,
    X=x_train,
    y=y_train,
    scoring='f1_macro',
    cv=10,
    n_jobs=-1
)
f1_average = f1_scores.mean()


print("F1 average score:", f1_average)


GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('tfidf',
                                        TfidfVectorizer(tokenizer=<function tokenizer_vect at 0x000001DCE73C68B0>)),
                                       ('clf', LinearSVC())]),
             n_jobs=-1,
             param_grid={'clf__C': (0.1, 1, 10), 'clf__dual': (True, False),
                         'clf__loss': ('hinge', 'squared_hinge'),
                         'clf__penalty': ('l1', 'l2'),
                         'tfidf__max_df': (0.5, 0.75, 1.0),
                         'tfidf__min_df': (1, 2),
                         'tfidf__ngram_range': [(1, 1), (1, 2)],
                         'tfidf__use_idf': (True, False)},
             scoring='f1_macro')

F1 average score: 0.709106802861895


<font size="4">3. DecisionTreeClassifier:</font>

In [19]:
param_grid = {
    'clf__criterion': ['gini', 'entropy'],
    'clf__max_depth': [None, 5, 10],
    'clf__min_samples_split': [2, 5, 10]
}

# Define the pipeline
pipeline = Pipeline([
    ('vect', TfidfVectorizer(min_df=5, ngram_range=(1, 6))),
    ('clf', DecisionTreeClassifier(random_state=42))
])

#Perform grid search with cross-validation using the pipeline
grid_search = GridSearchCV(
    pipeline,
    param_grid,
    cv=10,
    scoring='f1_macro',  # Using F1-score with macro averaging directly
    n_jobs=-1
)
grid_search.fit(x_train, y_train)

DecisionTree_model = grid_search.best_estimator_

#Calculate F1-score
f1_scores = cross_val_score(
    estimator=grid_search.best_estimator_,
    X=x_train,
    y=y_train,
    scoring='f1_macro',  # Using F1-score with macro averaging directly
    cv=10,
    n_jobs=-1
)
f1_average = f1_scores.mean()

#Print the F1-score average
print("F1 average score:", f1_average)



GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('vect',
                                        TfidfVectorizer(min_df=5,
                                                        ngram_range=(1, 6))),
                                       ('clf',
                                        DecisionTreeClassifier(random_state=42))]),
             n_jobs=-1,
             param_grid={'clf__criterion': ['gini', 'entropy'],
                         'clf__max_depth': [None, 5, 10],
                         'clf__min_samples_split': [2, 5, 10]},
             scoring='f1_macro')

F1-average score: 0.6241829745087971


<font size="4">Final results:</font>


Best F1 score : 0.709 - LinearSVC

Predict on the test set

In [21]:
y_pred = linearSVC_model.predict(x_test) # predicting the author's gender
predictions = df_test.copy()

for i in range(0,len(predictions['test_example_id'])):
    predictions['test_example_id'][i] = y_pred[i] #inserting the predictions

In [22]:
predictions

test_example_id                                              story
0                 m  כל קיץ אני והמשפחה נוסעים לארצות הברית לוס אנג...
1                 m  הגעתי לשירות המדינה אחרי שנתיים כפעיל בתנועת "...
2                 m  אחת האהבות הגדולות שלי אלו הכלבים שלי ושל אישת...
3                 m  רגע הגיוס לצבא היה הרגע הכי משמעותי עבורי, אני...
4                 f  אני הגעתי לברזיל ישר מקולומביה וגם אני עשיתי ע...
..              ...                                                ...
318               m  בשנה האחרונה הרגשתי די תקוע בעבודה, השגרה הפכה...
319               m  אני ואילן חברים טובים מזה 20 שנה תמיד חלמנו לפ...
320               m  מידי יום שישי אני נוהג לנסוע בתחבורה ציבורית ס...
321               m  לפני מספר חודשים, בשיא התחלואה של הגל השני, עמ...
322               m  היום בו דיווחתי על גניבה של האוטו שלי. בוקר אח...

[323 rows x 2 columns]

test_example_id predicted_category
0                  0                  m
1                  1                  m
2                  2                  m
3                  3                  m
4                  4                  f
..               ...                ...
318              318                  m
319              319                  m
320              320                  m
321              321                  m
322              322                  m

[323 rows x 2 columns]

### Save output to csv (optional)
After you're done save your output to the 'classification_results.csv' csv file.<br/>
We assume that the dataframe with your results contain the following columns:
* column 1 (left column): 'test_example_id'  - the same id associated to each of the test stories to be predicted.
* column 2 (right column): 'predicted_category' - the predicted gender value for each of the associated story. 

Assuming your predicted values are in the `df_predicted` dataframe, you should save you're results as following:

In [ ]:
df_predicted.to_csv('classification_results.csv',index=False)